### PR TITLE
Vet clinic database: add "join table" for visits

### PR DESCRIPTION
**Pull request by Feliverse**
### Learning objectives

- Understand the different types of relationships between tables.
- Prepare complex queries that answer analytical questions.
- Use primary key & foreign key mechanisms for joining tables.

Project requirements

- Create a table named vets
- Create a "join table" called specializations 
- Create a "join table" called visits
- Create a "join table" 
### Queries to answer the following:

- Who was the last animal seen by William Tatcher?
image

-  How many different animals did Stephanie Mendez see?
 image

- List all vets and their specialties, including vets with no specialties.
image

- List all animals that visited Stephanie Mendez between April 1st and August 30th, 2020.
image

- What animal has the most visits to vets?
image

- Who was Maisy Smith's first visit?
image

- Details for most recent visit: animal information, vet information, and date of visit.
image

- How many visits were with a vet that did not specialize in that animal's species?
image

- What specialty should Maisy Smith consider getting? Look for the species she gets the most.
image

### CODE REVIEW

File: schema.sql [hunk 1]
```diff
@@ -41,25 +41,21 @@ ALTER TABLE animals ADD COLUMN owner_id INT,
...
     vets_id INT,
   CONSTRAINT fk_species FOREIGN KEY(species_id) REFERENCES species(id) ON DELETE CASCADE,
-  CONSTRAINT fk_vets FOREIGN KEY(vets_id) REFERENCES vets(id) ON DELETE CASCADE
-);
+  CONSTRAINT fk_vets FOEIGN KEY(vets_id) REFERENCES vets(id) ON DELETE CASCADE);
 
 CREATE TABLE visits (
     animals_id INT,
     vets_id INT,
   date DATE,
```
Comment (Deepakdanger) [inline schema.sql]: - [x] Here using `date` as the column is not recognized which date. Kindly use `visit_date DATE`.

File: data.sql [hunk 2]
```diff
@@ -31,12 +31,11 @@ UPDATE animals SET owner_id=4 WHERE name IN ('Charmander','Squirtle','Blossom');
...
-('Stephanie Mendez', 64, '1981-05-04'),
-('Jack Harkness', 38, '2008-06-08');
+  VALUES ('William Tatcher', 45, '2000-04-23'),
+  ('Maisy Smith', 26, '2019-01-17'),
+  ('Stephanie Mendez', 64, '1981-05-04'),
+  ('Jack Harkness', 38, '2008-06-08');
 
 INSERT INTO specializations (vets_id,species_id) 
-VALUES (1,1),(3,2),(3,1),(4,2);
+  VALUES (1,1),(3,2),(3,1),(4,2);
```
Comment (Deepakdanger) [inline data.sql:41]: - [x] Here we have to Insert data into visits table also .`visits` table (**students can use different queries than in the example, for example they can use hardcoded animal_id and vet_id**)
 ```
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Agumon'), (SELECT id FROM vets WHERE name = 'William Tatcher'), '2020-05-24');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Agumon'), (SELECT id FROM vets WHERE name = 'Stephanie Mendez'), '2020-07-22');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Gabumon'), (SELECT id FROM vets WHERE name = 'Jack Harkness'), '2021-02-02');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Pikachu'), (SELECT id FROM vets WHERE name = 'Maisy Smith'), '2020-01-05');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Pikachu'), (SELECT id FROM vets WHERE name = 'Maisy Smith'), '2020-03-08');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Pikachu'), (SELECT id FROM vets WHERE name = 'Maisy Smith'), '2020-05-14');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Devimon'), (SELECT id FROM vets WHERE name = 'Stephanie Mendez'), '2021-05-04');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Charmander'), (SELECT id FROM vets WHERE name = 'Jack Harkness'), '2021-02-24');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Plantmon'), (SELECT id FROM vets WHERE name = 'Maisy Smith'), '2019-12-21');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Plantmon'), (SELECT id FROM vets WHERE name = 'William Tatcher'), '2020-08-10');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Plantmon'), (SELECT id FROM vets WHERE name = 'Maisy Smith'), '2021-04-07');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Squirtle'), (SELECT id FROM vets WHERE name = 'Stephanie Mendez'), '2019-09-29');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Angemon'), (SELECT id FROM vets WHERE name = 'Jack Harkness'), '2020-10-03');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Angemon'), (SELECT id FROM vets WHERE name = 'Jack Harkness'), '2020-11-04');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Boarmon'), (SELECT id FROM vets WHERE name = 'Maisy Smith'), '2019-01-24');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Boarmon'), (SELECT id FROM vets WHERE name = 'Maisy Smith'), '2019-05-15');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Boarmon'), (SELECT id FROM vets WHERE name = 'Maisy Smith'), '2020-02-27');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Boarmon'), (SELECT id FROM vets WHERE name = 'Maisy Smith'), '2020-08-03');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Blossom'), (SELECT id FROM vets WHERE name = 'Stephanie Mendez'), '2020-05-24');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Blossom'), (SELECT id FROM vets WHERE name = 'William Tatcher'), '2021-01-11');
```

File: queries.sql [hunk 3]
```diff
@@ -149,4 +106,4 @@ LEFT JOIN specializations s ON s.vets_id=vets.id
 JOIN animals ON animals.species_id=s.species_id 
 JOIN species ON species.id=s.species_id 
 WHERE vets.name='Maisy Smith' 
-GROUP BY species.name ORDER BY count_species DESC LIMIT 1;
\ No newline at end of file
+GROUP BY species.name ORDER BY count_species DESC LIMIT 1;
```
Comment (Deepakdanger) [inline queries.sql]: - [x] Here for
-  * How many times visits where with a vet that did not specialize in that animal's species?

```sql
  SELECT count(*)
  FROM visits
  LEFT JOIN animals ON animals.id = visits.animal_id
  LEFT JOIN vets ON vets.id = visits.vet_id
  WHERE animals.species_id NOT IN (SELECT species_id FROM specializations WHERE vet_id = vets.id)
```

- What specialty should Maisy Smith consider getting? Look for the species she gets the most.

```sql
  SELECT species.name, count(*)
  FROM visits
  LEFT JOIN animals ON animals.id = visits.animal_id
  LEFT JOIN species ON animals.species_id = species.id
  LEFT JOIN vets ON vets.id = visits.vet_id
  WHERE vet.name = 'Maisy Smith'
  GROUP BY species.name
```

File: queries.sql [hunk 4]
```diff
@@ -77,50 +77,7 @@ SELECT animals.name FROM animals JOIN owners ON animals.owner_id=owners.id WHERE
...
-  SELECT MAX(myf.count_visits) 
-  FROM ( select a.name, COUNT(a.name) 
-  AS count_visits from visits v 
-  JOIN animals a ON v.animals_id=a.id 
-  JOIN vets vt ON v.vets_id=vt.id 
-  GROUP BY a.name) 
-  AS myf
-);
-
+/*Queries in join tables*/
```
Comment (Deepakdanger) [inline queries.sql:80]: - [ ] Here also we have to write query for many questions:-
- [ ] Who was the last animal seen by William Tatcher?
- [ ] How many different animals did Stephanie Mendez see?
- [ ] List all vets and their specialties, including vets with no specialties.
- [ ] List all animals that visited Stephanie Mendez between April 1st and August 30th, 2020
- [ ] What animal has the most visits to vets?

**Review by Deepakdanger — CHANGES_REQUESTED** [review]
Hi @Feliverse , 

Good job so far! 👏
There are some issues that you still need to work on to go to the next project but you are almost there!

## To Highlight :100: 
- No linter error :heavy_check_mark: 
- Website looks professional :+1: 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: schema.sql (hunk 1 above)
Comment (Feliverse) [reply schema.sql]: That is a great solution, thank you

File: queries.sql (hunk 3 above)
Comment (Feliverse) [reply queries.sql]: Thank you, this is the correct solution. I was don't understanding the question.

File: queries.sql (hunk 4 above)
Comment (Feliverse) [reply queries.sql:80]: Ups, I will sent the final commit whit all this queries.

File: queries.sql (hunk 3 above)
Comment (Feliverse) [reply queries.sql]: I am adding solutions in the PR description

File: queries.sql (hunk 4 above)
Comment (Abenezer-Tilahun) [reply queries.sql:80]: Fixed.

File: schema.sql (hunk 1 above)
Comment (Abenezer-Tilahun) [reply schema.sql]: Fixed.

File: queries.sql (hunk 3 above)
Comment (Abenezer-Tilahun) [reply queries.sql]: Fixed.

**Review by Abenezer-Tilahun — APPROVED** [review]
Hi @Feliverse 
 
Your project is complete  👍 💯   There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 👏🏻

✔️ No linter errors.
✔️ You used correct Git-Flow.
✔️ Professional README  
✔️ Good Commit Messages
✔️ Good job for Implemented Suggested changes.

## Optional suggestions 

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**